### PR TITLE
SMK-1220 / Added new functionalities

### PR DESCRIPTION
**Pull request by Highvolt**
* Updated to last version of bot framework.
* Added validations as middleware
* Added Value Parsers as middleware
* Added possibility to node override the next node to be executed
* Added validation that used external functions (using nodejs sandboxes see: https://nodejs.org/api/vm.html)
* Added validation for response length
* Added parsers for date, time and email (this last one is mandatory, because bot framework is not consistent across all connectors)
* Support to handle an carrousel item selection

### CODE REVIEW

**Comment by Highvolt** [comment]
Please release version 3.5

File: lib/GraphDialog.js [hunk 1]
```diff
@@ -1,145 +1,164 @@
...
-        var _this = this;
-        return new Promise(function (resolve, reject) {
-            _this.parser = new Parser_1.Parser(_this.options);
-            _this.parser.init().then(function () {
+    }
+    init() {
+        return new Promise((resolve, reject) => {
+            this.parser = new Parser_1.Parser(this.options);
+            this.parser.init().then(() => {
                 console.log('parser is ready');
```
Comment (ruicampos) [inline lib/GraphDialog.js:45]: shouldn't we have a 'debug' mode and log only when in debugging?

File: lib/GraphDialog.js [hunk 2]
```diff
@@ -1,145 +1,164 @@
...
             }
         }
         ;
         session.cancelDialog(dialogIndex, this.internalPath);
-    };
-    GraphDialog.prototype.getDialog = function () {
-        var _this = this;
+    }
+    getDialog() {
         console.log('get dialog');
```
Comment (ruicampos) [inline lib/GraphDialog.js:78]: Same point about logging.

File: lib/GraphDialog.js [hunk 3]
```diff
@@ -1,145 +1,164 @@
...
+                    return this.options.onAfterProcessingStep.call(this, session, args, next);
                 else
                     return next();
             },
-            function (session, args, next) {
+            (session, args, next) => {
+                return this.setNextStepHandler(session, args, next);
+            },
+            (session, args, next) => {
                 console.log('calling loop function');
```
Comment (ruicampos) [inline lib/GraphDialog.js:115]: logging point...

File: lib/GraphDialog.js [hunk 4]
```diff
@@ -1,145 +1,164 @@
...
-    };
-    GraphDialog.prototype.stepInteractionHandler = function (session, results, next) {
-        var _this = this;
+    }
+    stepInteractionHandler(session, results, next) {
         session.privateConversationData._lastMessage = session.message && session.message.text;
-        var currentNode = this.nav.getCurrentNode(session);
-        console.log("perform action: " + currentNode.id + ", " + currentNode.type);
+        let currentNode = this.nav.getCurrentNode(session);
+        console.log(`perform action: ${currentNode.id}, ${currentNode.type}`);
```
Comment (ruicampos) [inline lib/GraphDialog.js:123]: logging point...

**Review by ruicampos — CHANGES_REQUESTED** [review]
Only point is about logging. As we discussed we can add a logger with the logging level based on an environment variable - default: INFO (if environment variable not found)

**Review by ruicampos — APPROVED** [review]
I will merge the PR and logging can be changed in next issue.

**Pull request merged**